### DISCUSSION TITLE
Clock drift / Block number considerations

### DISCUSSION THREAD

**Discussion by frangio** [pooltogether/multichain-gov-proposal-spec#3] [General]
The security section mentions clock drift as a concern, but a bigger concern in my mind is the fact that Governor is based on block numbers. (Some context: For OpenZeppelin Governor while we identified this as an undesirable feature, the decision was made to remain compatible with Compound's contracts and existing infrastructure. We did not see at the time that it would be an even bigger issue for multi-chain settings.)

By itself this is likely an issue: on some chains block numbers may be totally infeasible as a measure of time, as the concept of blocks gets blurred in many of them and they may not have a reliable average block time as Ethereum does. If the implementation will be Governor-based, this is an important thing to look into, and to find ways around it. See https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3080 for some prior discussion.

If this somehow turns out not to be an issue and the Governor can remain block number based, not clock drift but "block number drift" can for sure be an issue, as in general one chain can only make a rough assumption of the block number another chain is at.

Overall moving towards timestamps would be much better overall, but there are some concerns discussed in the PR linked above.

**Comment by asselstine** [comment]
Thank for that link! Good to see that discussion; as I think the block number Timer is one of the core issues making Governor less reusable.  After reading the [Rollcall](https://github.com/withtally/rollcall) code, it seemed that the block number Timers were the main reason for forking the entire Governor contract.

If the Governor is updated we'll be able to re-use more code, but if it isn't I bet we'll still be able to bolt-on a compatibility layer.

Really, the only difference is the start + end timestamps and some custom logic. We could create a contract that inherits from both Governor and GovernorRoot.  It will use a little more storage (for both types of timestamps) and will have to manually tune the length in blocks and seconds to match, but it's totally possible.
